rust/kernel/mem_cache.rs

```diff
@@ -0,0 +1,58 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! Kernel memory caches (kmem_cache).
+//!
+//! C headers: [`include/linux/slab.h`](../../include/linux/slab.h)
+
+use crate::error::{code::*, Result};
+use crate::{bindings, str::CStr};
+use core::ptr;
+
+/// A kernel memory cache.
+///
+/// This isn't ready to be made public yet because it only provides functionality useful for the
+/// allocation of inodes in file systems.
+pub(crate) struct MemCache {
+    ptr: ptr::NonNull<bindings::kmem_cache>,
+}
+
+impl MemCache {
+    #[allow(dead_code)]
+    pub(crate) fn try_new<T>(
+        name: &'static CStr,
+        init: Option<unsafe extern "C" fn(*mut core::ffi::c_void)>,
+    ) -> Result<Self> {
```

Comment:
I'll add docs.

As for making this function `unsafe`, I'm not sure it's needed: although `init` is `unsafe`, we still require an `unsafe` block within it to do anything that is unsafe.

My usual test for marking a function `unsafe` is: does it allow callers to get UB without `unsafe` blocks? I don't think it's the case here. (Though I'd be happy to be shown otherwise.)

Reply:
I don't think we want the callee to reason about who's caller. It should only reason about its safety requirement, which is much more than what an unsafe keyword on fn signature can convey.

So I think this function should be unsafe since it's try_new's caller's responsibility to ensure that the safety requirement matches.

Reply:
It's called when an entry in the cache is allocated. It's meant to initialise the entry.

Reply:
So it is an initializer for type `T`? I think using an unsafe extern fn pointer doesnt really fit then.
- Why is it optional? What default would the C side know to use for custom Rust types?
- Why not use `fn() -> T` or if you need in-place/pinned init `fn() -> impl [Pin]Init<T>`?

Reply:
It's a bit more complicated. It's called when entries are _allocated_ into the cache and only once, so the sequence is: a "slab" is allocated, entries in it are "initialised-once" via this this call (e.g., see https://elixir.bootlin.com/linux/latest/source/fs/inode.c#L416), then when the entry is actually allocated for use, another per-use init is performed (this is the traditional init, but assumes that init-once was called before). Eventually, when the entry is dropped, some form of destructor is called (but leaves the init-once state consistent) and the entry goes back to free state; when it's allocated again (if ever), init-once doesn't have to run again, just the regular init runs, and so on.

The idea is to avoid repeating work for fields that remain consistent. We don't really use this concept in Rust, but the C side for inodes does, so we need to accommodate for that.

Having said all of this, this is not a public API for Rust modules. The _only_ reason I wrote this was to automatically destroy it when it goes out of scope during pin-init and I wasn't sure if scope guards would work as expected. I am also not interested in developing this abstraction any further than I need to for this. Perhaps this will change when it's actually needed by some module.

Reply:
I see, then I would just hardcode that usage and not take the `Option<fn>` parameter at all. That way you do not need any safety requirements on that function.